include/smack/smack.h

```diff
@@ -135,32 +287,146 @@ void __SMACK_decls() {
+  D("function {:inline} $load.i8(M:[ref]i8, p:ref) returns (i8){M[p]}");
+
+  D("function {:inline} $store.i64(M:[ref]i8, p:ref, v:i64) returns ([ref]i8)"
+    "{M[p := v[8:0]][$add.i32(p, 1bv32) := v[16:8]][$add.i32(p, 2bv32) := v[24:16]][$add.i32(p, 3bv32) := v[32:24]]"
+    "[$add.i32(p, 4bv32) := v[40:32]][$add.i32(p, 5bv32) := v[48:40]][$add.i32(p, 6bv32) := v[56:48]][$add.i32(p, 7bv32) := v[64:56]]}");
+  D("function {:inline} $store.i32(M:[ref]i8, p:ref, v:i32) returns ([ref]i8) {M[p := v[8:0]][$add.i32(p, 1bv32) := v[16:8]][$add.i32(p, 2bv32) := v[24:16]][$add.i32(p, 3bv32) := v[32:24]]}");
+  D("function {:inline} $store.i16(M:[ref]i8, p:ref, v:i16) returns ([ref]i8) {M[p := v[8:0]][$add.i32(p, 1bv32) := v[16:8]]}");
+  D("function {:inline} $store.i8(M:[ref]i8, p:ref, v:i8) returns ([ref]i8) {M[p := v]}");
+  D("function {:inline} $isExternal(p: ref) returns (bool) { $slt.ref(p, $sub.ref($GLOBALS_BOTTOM, 32768bv32)) }");
+#endif
+  D("function {:inline} $trunc.i64.i32(p: i64) returns (i32) {p[32:0]}");
+  D("function {:inline} $trunc.i64.i16(p: i64) returns (i16) {p[16:0]}");
+  D("function {:inline} $trunc.i64.i8(p: i64) returns (i8) {p[8:0]}");
+  D("function {:inline} $trunc.i32.i16(p: i32) returns (i16) {p[16:0]}");
+  D("function {:inline} $trunc.i32.i8(p: i32) returns (i8) {p[8:0]}");
+  D("function {:inline} $trunc.i16.i8(p: i16) returns (i8) {p[8:0]}");
+
+  D("function {:inline} $zext.i8.i64(p: i8) returns (i64) {(0bv56)++p}");
+  D("function {:inline} $zext.i8.i32(p: i8) returns (i32) {(0bv24)++p}");
+  D("function {:inline} $zext.i8.i16(p: i8) returns (i16) {(0bv8)++p}");
+  D("function {:inline} $zext.i16.i64(p: i16) returns (i64) {(0bv48)++p}");
+  D("function {:inline} $zext.i16.i32(p: i16) returns (i32) {(0bv16)++p}");
+  D("function {:inline} $zext.i32.i64(p: i32) returns (i64) {(0bv32)++p}");
+
+  D("function {:inline} $sext.i8.i64(p: i8) returns (i64) {if $sge.i8(p, 0bv8) then $zext.i8.i64(p) else (($neg.i64(1bv64))[64:8])++p}");
+  D("function {:inline} $sext.i8.i32(p: i8) returns (i32) {if $sge.i8(p, 0bv8) then $zext.i8.i32(p) else (($neg.i32(1bv32))[32:8])++p}");
+  D("function {:inline} $sext.i8.i16(p: i8) returns (i16) {if $sge.i8(p, 0bv8) then $zext.i8.i16(p) else $neg.i8(1bv8)++p}");
+  D("function {:inline} $sext.i16.i64(p: i16) returns (i64) {if $sge.i16(p, 0bv16) then $zext.i16.i64(p) else $neg.i64(1bv64)[64:16]++p}");
+  D("function {:inline} $sext.i16.i32(p: i16) returns (i32) {if $sge.i16(p, 0bv16) then $zext.i16.i32(p) else $neg.i16(1bv16)++p}");
+  D("function {:inline} $sext.i32.i64(p: i32) returns (i64) {if $sge.i32(p, 0bv32) then $zext.i32.i64(p) else $neg.i32(1bv32)++p}");
+
+  D("function {:inline} $p2i(p: ref) returns (size) {p}");
+  D("function {:inline} $i2p(p: size) returns (ref) {p}");
+  D("function {:inline} $p2b(p: ref) returns (bool) {p != $NULL}");
+#else
+  D("axiom $NULL == 0;");
   D("function {:inline} $pa(pointer: int, index: int, size: int) returns (int) {pointer + index * size}");
```

Comment:
This, and the following definitions, should use `ref` and `size` rather than `int`.